### ISSUE TITLE
question about edge feature encoding

### ISSUE THREAD

**Issue by YJiangcm** [microsoft/Graphormer#115]
I notice that in **graphormer/modules/graphormer_layers.py** you modify the spatial_pos_ by setting 1 to 1, x > 1 to x - 1, could you please tell me the reason?

        if self.edge_type == "multi_hop":
            spatial_pos_ = spatial_pos.clone()
            spatial_pos_[spatial_pos_ == 0] = 1  # set pad to 1
            # set 1 to 1, x > 1 to x - 1
            spatial_pos_ = torch.where(spatial_pos_ > 1, spatial_pos_ - 1, spatial_pos_)

**Comment by mavisguan** [comment]
Hi, sorry for the late response. (Other authors have been too busy to answer)
- About setting x > 1 to x - 1: in function `pad_spatial_pos_unsqueeze()` in collator.py because we increment each element in `spatial_pos` by 1, and add zero padding. So later we set x > 1 to x - 1 to restore spatial_pos to real shortest distance between nodes.
- About setting 1 to 1: we calculate an average value of edge embedding along the shortest path just a bit after your code snippet. If we set 1 to 0, there'll be division by zero, so we set 1 to 1.
```python
edge_input = (
                edge_input.sum(-2) / (spatial_pos_.float().unsqueeze(-1))
            ).permute(0, 3, 1, 2)
```

**Comment by mavisguan** [comment]
> I notice that in **graphormer/modules/graphormer_layers.py** you modify the spatial_pos_ by setting 1 to 1, x > 1 to x - 1, could you please tell me the reason?
> 
> ```
>     if self.edge_type == "multi_hop":
>         spatial_pos_ = spatial_pos.clone()
>         spatial_pos_[spatial_pos_ == 0] = 1  # set pad to 1
>         # set 1 to 1, x > 1 to x - 1
>         spatial_pos_ = torch.where(spatial_pos_ > 1, spatial_pos_ - 1, spatial_pos_)
> ```

Hi, sorry for the late response. (Other team members have been just too busy to answer!)

About setting x > 1 to x - 1: in function pad_spatial_pos_unsqueeze() in collator.py because we increment each element in spatial_pos by 1, and add zero padding. So later we set x > 1 to x - 1 to restore spatial_pos to real shortest distance between nodes.
About setting 1 to 1: we calculate an average value of edge embedding along the shortest path just a bit after your code snippet. If we set 1 to 0, there'll be division by zero, so we set 1 to 1.
edge_input = (
                edge_input.sum(-2) / (spatial_pos_.float().unsqueeze(-1))
            ).permute(0, 3, 1, 2)

**Comment by YJiangcm** [comment]
Thanks for you answering : )
